tests/lib/external/prepare-ssh.sh

```diff
@@ -3,17 +3,38 @@ set -ex
 
 INSTANCE_IP="${1:-localhost}"
 INSTANCE_PORT="${2:-8022}"
-USER="${3:-$(whoami)}"
+USER_NAME="${3:-$(whoami)}"
+USER_PASS="${4:-}"
+USER_TYPE="${5:-}"
 
 execute_remote(){
     # shellcheck disable=SC2029
-    ssh -q -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -p "$INSTANCE_PORT" "$USER@$INSTANCE_IP" "$@"
+    $SSHPASS_PARAM ssh -q -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -p "$INSTANCE_PORT" "$USER_NAME@$INSTANCE_IP" "$@"
 }
 
-execute_remote "sudo adduser --uid 12345 --extrausers --quiet --disabled-password --gecos '' test"
-execute_remote "echo test:ubuntu | sudo chpasswd"
-execute_remote "echo 'test ALL=(ALL) NOPASSWD:ALL' | sudo tee /etc/sudoers.d/create-user-test"
+USER_TYPE_PARAM=""
+if [ -z "$USER_TYPE" ]; then
+    USER_TYPE_PARAM="--extrausers"
+fi
 
-execute_remote "sudo adduser --extrausers --quiet --disabled-password --gecos '' external"
-execute_remote "echo external:ubuntu | sudo chpasswd"
-execute_remote "echo 'external ALL=(ALL) NOPASSWD:ALL' | sudo tee /etc/sudoers.d/create-user-external"
+# It is needed to use sshpass to connect using password
+SSHPASS_PARAM=""
+if [ -n "$USER_PASS" ]; then
+    SSHPASS_PARAM="sshpass -p $USER_PASS"
+fi
+
+# When user is root, neither sudo nor sshpass is needed
+SUDO_PARAM="sudo"
+if [ "$USER_NAME" = "root" ]; then
+    SUDO_PARAM=""
+elif [ -n "$USER_PASS" ]; then
+	SUDO_PARAM="sshpass -p $USER_PASS sudo"
+fi
```

Comment:
this whole logic is confusing to me...

If we have a USER_PASS defined, then above, SSHPASS_PARAM gets set to `"sshpass -p $USER_PASS"`, but then if USER_NAME is _not_ root, then we will also define SUDO_PARAM to "sshpass -p $USER_PASS sudo", which when we run the first command, 

```
execute_remote "$SUDO_PARAM adduser ..."
```

will now get expanded to
```
execute_remote "sshpass -p $USER_PASS sudo adduser ..."
```

and due to defining SSHPASS_PARAM above and using it in the definition of execute_remote we will have the following command:

```
$SSHPASS_PARAM ssh ... "$@"
```

which then expands to the erroneous command:

```
"sshpass -p $USER_PASS" ssh ... sshpass -p $USER_PASS sudo adduser ..."
```

which is wrong I think

can you clarify more what you were intending these variables to be used for in this script?

Reply:
iirc it is intentional, first, we use sshpass to run the ssh and then we use sshpass again to run a sudo adduser command
@anonymouse64 does it make sense?

Reply:
ah I see now, so sshpass really is just a little wrapper program which can spit out a password as output to a program like `sudo` or `ssh`, this makes sense to me now then